### PR TITLE
Add less-boilerplate way to override thread context class loader

### PR DESCRIPTION
**Pull request by sergey-melnychuk**
Avoid C++-like RAII approach in favour of more Java-like lambda-friendly approach.

Just 

```
withClassLoader(classLoader, () -> doSomething());
```

instead of 

```
try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader) {
        doSomething();
}
```

`Runnable`, `Supplier<T>` and custom `ThrowingSupplier<T, E extends Exception>` are supported.

### CODE REVIEW

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 1]
```diff
@@ -13,22 +13,59 @@
...
+    public interface ThrowingSupplier<T, E extends Exception>
+            extends Supplier<T>
+    {
+        default T get()
+        {
+            try {
+                return supplyOrThrow();
+            }
+            catch (Exception e) {
+                // TODO(sergey-melnychuk) There must be a better way to address this
```
Comment (sergey-melnychuk) [inline core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:64]: Not sure how exactly, but I have a strong feeling that this can be done better than just wrapping into `RuntimeException`. Suggestions?

**Comment by kokosing** [comment]
> Avoid C++-like RAII approach 

What is wrong with this?

**Comment by sergey-melnychuk** [comment]
> > Avoid C++-like RAII approach
> 
> What is wrong with this?

What I don't like is that the only correct usage of `ThreadContextClassLoader` requires `try`-block, that introduces unnecessary complication and is potentially error-prone. Why create an object that is **always** ignored? Whereas with function call there is no way to misuse it.

There is nothing wrong with RAII though, but IMO it fits C++ much better than Java due to lack of destructor concept in Java.

This is only a suggestion, not a demand :) Feel free to close the PR if it doesn't do as much good as I thought.

File: core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java [hunk 2]
```diff
@@ -165,12 +165,14 @@ public synchronized void stop()
 
         for (Map.Entry<CatalogName, MaterializedConnector> entry : connectors.entrySet()) {
             Connector connector = entry.getValue().getConnector();
-            try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(connector.getClass().getClassLoader())) {
-                connector.shutdown();
-            }
-            catch (Throwable t) {
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java:171]: Previously this was also catching exceptions from `ThreadContextClassLoader.close()`

**Review by kokosing — COMMENTED** [review]
Thanks for explanation.

File: core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java (hunk 2 above)
Comment (sergey-melnychuk) [reply core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java:171]: moved inner try-block into outer one

**Comment by wendigo** [comment]
I like this change :) Less boilerplate and more opinionated way of using threadcontextloader.

**Comment by sergey-melnychuk** [comment]
Still got some more suspicious failing tests to investigate.

**Comment by sergey-melnychuk** [comment]
CI is green: https://github.com/trinodb/trino/actions/runs/1053179897

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java [hunk 3]
```diff
@@ -98,13 +98,15 @@ void invalidateCache()
...
     public Principal createAuthenticatedPrincipal(String user, String password)
     {
-        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
-            return authenticationCache.getUnchecked(new Credential(user, password));
-        }
-        catch (UncheckedExecutionException e) {
-            throwIfInstanceOf(e.getCause(), AccessDeniedException.class);
-            throw e;
-        }
+        return withClassLoader(getClass().getClassLoader(), () -> {
```
Comment (wendigo) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java:101]: we could have and overload that accepts Class<?> and uses it's `getClassLoader()` (instead of chaining calls)

Comment (findepi) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java:101]: what would be its signature?

**Comment by findepi** [comment]
@sergey-melnychuk was there any case with checked exceptions?

**Comment by sergey-melnychuk** [comment]
@findepi only `ClassLoaderSafeSchemaRegistryClient`, as Confluent's `throws IOException, RestClientException` has `throws IOException, RestClientException` pretty much on every method.

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 4]
```diff
@@ -13,22 +13,55 @@
...
+        default T get()
+        {
+            try {
+                return supplyOrThrow();
+            }
+            catch (RuntimeException e) {
+                throw e;
+            }
+            catch (Exception e) {
+                throw new RuntimeException(e);
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:61]: We should have ThrowingSupplier that just throws, so that exceptions are propagated as-is,

Once you do that `E extends Exception` -> `E extends Throwable`.

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 5]
```diff
@@ -13,22 +13,55 @@
...
-        Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        withClassLoader(newThreadContextClassLoader, () -> {
+            runnable.run();
+            return null;
+        });
+    }
+
+    public static <T, E extends Exception> T withClassLoader(ClassLoader newThreadContextClassLoader, ThrowingSupplier<T, E> supplier)
+    {
+        return withClassLoader(newThreadContextClassLoader, (Supplier<T>) supplier);
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:44]: i would delegate the other way around.

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 6]
```diff
@@ -13,22 +13,55 @@
...
+    }
+
+    @FunctionalInterface
+    public interface ThrowingSupplier<T, E extends Exception>
+            extends Supplier<T>
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:49]: This won't work for:

> as Confluent's SchemaRegistryClient has throws IOException, RestClientException pretty much on every method.

since the two exception types are not in direct hierarchy

we would need to accept E1, E2... would that work?

Comment (sergey-melnychuk) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:49]: If this doesn't work - then `ClassLoaderSafeSchemaRegistryClient` is not covered by any functional tests in the CI :) 

The way I understand it, `... throws E1 extends Exception, E2 extends Exception` is reduced to `... throws Exception` which kind of makes sense as each E1 and E2 'is an' Exception. I believe the compiler interprets it in a similar way:

[JLS 18.2.5 Checked Exception Constraints](https://docs.oracle.com/javase/specs/jls/se11/html/jls-18.html#jls-18.2.5)

>  Otherwise, let E1, ..., En be the types in the function type's throws clause that are not proper types. If the lambda expression is implicitly typed, let its parameter types be the function type's parameter types. If the lambda body is a poly expression or a block containing a poly result expression, let the targeted return type be the function type's return type. Let X1, ..., Xm be the checked exception types that the lambda body can throw (§11.2). Then there are two cases:

> If n = 0 (the function type's throws clause consists only of proper types), then if there exists some i (1 ≤ i ≤ m) such that Xi is not a subtype of any proper type in the throws clause, the constraint reduces to false; otherwise, the constraint reduces to true.

> If n > 0, the constraint reduces to a set of subtyping constraints: for all i (1 ≤ i ≤ m), if Xi is not a subtype of any proper type in the throws clause, then the constraints include, for all j (1 ≤ j ≤ n), ‹Xi <: Ej›. In addition, for all j (1 ≤ j ≤ n), the constraint reduces to the bound throws Ej.

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java (hunk 5 above)
Comment (sergey-melnychuk) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:44]: I was thinking of delegation from more-specific `ThrowingSupplier<T, E>`-version to more-generic `Supplier<T>`-version to keep `Supplier<T>`-version unaware of any checked exceptions and to point to proper `Supplier::get`, which is overridden in `ThrowingSupplier`.

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java (hunk 3 above)
Comment (sergey-melnychuk) [reply plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java]: added in https://github.com/trinodb/trino/pull/8507/commits/a5235f11dd4b824ce18d7dd6cc9025e36a98505d, PTAL

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java (hunk 4 above)
Comment (sergey-melnychuk) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:76]: It leads to another round of "Checked vs. Unchecked" fight:

Screenshot 2021-07-22 at 11 29 03

Comment (findepi) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:76]: `ThrowingSupplier` must not be a `Supplier`  (as it is on your screenshot)

Comment (sergey-melnychuk) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java]: makes sense, PTAL: https://github.com/trinodb/trino/pull/8507/commits/e60dfc9d2b3e1fc80f1d3e4520670f890c8acf1d

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java (hunk 5 above)
Comment (sergey-melnychuk) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java]: done

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java (hunk 6 above)
Comment (sergey-melnychuk) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java]: `ThrowingSupplier` does not extend `Supplier` since https://github.com/trinodb/trino/pull/8507/commits/e60dfc9d2b3e1fc80f1d3e4520670f890c8acf1d, resolving this thread.

**Comment by wendigo** [comment]
@sergey-melnychuk please fix checkstyle issues (I guess around `catch` formatting in last commit). Other than that it looks great, thanks!

File: core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java [hunk 7]
```diff
@@ -337,12 +337,14 @@ private synchronized void removeConnectorInternal(CatalogName catalogName)
         MaterializedConnector materializedConnector = connectors.remove(catalogName);
         if (materializedConnector != null) {
             Connector connector = materializedConnector.getConnector();
-            try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(connector.getClass().getClassLoader())) {
-                connector.shutdown();
-            }
-            catch (Throwable t) {
-                log.error(t, "Error shutting down connector: %s", catalogName);
-            }
+            withClassLoader(connector.getClass().getClassLoader(), () -> {
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java:340]:
```suggestion
            withClassLoader(connector.getClass(), () -> {
```

Comment (sergey-melnychuk) [reply core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java]: thanks, fixed this and bunch of others in https://github.com/trinodb/trino/pull/8507/commits/b01458ea6fe70be739cf138df909969b10e08210

**Comment by sergey-melnychuk** [comment]
@wendigo thanks, done. Once CI on https://github.com/trinodb/trino/pull/8507/commits/b01458ea6fe70be739cf138df909969b10e08210 is green, I'll squash everything to single commit.

File: core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java [hunk 8]
```diff
@@ -165,8 +165,8 @@ public synchronized void stop()
 
         for (Map.Entry<CatalogName, MaterializedConnector> entry : connectors.entrySet()) {
             Connector connector = entry.getValue().getConnector();
-            try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(connector.getClass().getClassLoader())) {
-                connector.shutdown();
+            try {
+                withClassLoader(connector.getClass(), connector::shutdown);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java:169]: > `withClassLoader(connector.getClass(),`

this is wrong. a Class is not a ClassLoader.

Comment (sergey-melnychuk) [reply core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java:169]: This override makes chained `getClassLoader()` calls unnecessary and just takes class-loader for a class:

```
public static void withClassLoader(Class<?> clazz, Runnable runnable)
{
    withClassLoader(clazz.getClassLoader(), runnable);
}
```

https://github.com/trinodb/trino/pull/8507/files#diff-d5f1fa216ac7b2b536a6fa79582157d1270bdf985b6901056e5074c355be2daeR20

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java:169]: i know that it does that. But still, at the call site, it _reads wrong_.

Comment (sergey-melnychuk) [reply core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java]: Naming things is hard. It does read wrong, I get your point. Maybe 'withClassLoader**Of**(Class<?>, ...)'?

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java]: that would be better

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 9]
```diff
@@ -13,22 +13,49 @@
...
+
+    public static void withClassLoader(ClassLoader newThreadContextClassLoader, Runnable runnable)
+    {
+        withClassLoader(newThreadContextClassLoader, () -> {
+            runnable.run();
+            return null;
+        });
+    }
 
-    public ThreadContextClassLoader(ClassLoader newThreadContextClassLoader)
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:23]: This is SPI breaking change. Deprecate, but don't remove.

Or even don't deprecate, since the new version is _almost_ a drop-in replacement (multiple checked exceptions)

Comment (sergey-melnychuk) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java]: done, original implementation restored

**Review by wendigo — APPROVED** [review]
lgtm :)

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 10]
```diff
@@ -13,22 +13,49 @@
...
+        }
     }
 
-    @Override
-    public void close()
+    @FunctionalInterface
+    public interface ThrowingSupplier<T, E extends Exception>
     {
-        Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        T getOrThrow() throws E;
```
Comment (kokosing) [inline core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:74]: If checked is used then `orThrow()` is obvious. I would skip that.

File: core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java [hunk 11]
```diff
@@ -356,9 +358,8 @@ private Connector createConnector(CatalogName catalogName, InternalConnectorFact
                 pageIndexerFactory,
                 factory.getDuplicatePluginClassLoaderFactory());
 
-        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(factory.getConnectorFactory().getClass().getClassLoader())) {
-            return factory.getConnectorFactory().create(catalogName.getCatalogName(), properties, context);
-        }
+        return withClassLoader(factory.getConnectorFactory().getClass(),
+                () -> factory.getConnectorFactory().create(catalogName.getCatalogName(), properties, context));
```
[the remark this answers is not in the document]
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java:362]: withClassLoaderOf*

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 12]
```diff
@@ -13,22 +13,49 @@
  */
 package io.trino.spi.classloader;
 
-import java.io.Closeable;
-
 public class ThreadContextClassLoader
-        implements Closeable
```
[the remark this answers is not in the document]
Comment (sergey-melnychuk) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java]: Original implementation restored in 15f19b651d2f3a46f5ffd6f856c95f694adeb245

File: core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java [hunk 13]
```diff
@@ -112,10 +112,8 @@ private EventListener createEventListener(File configFile)
         EventListenerFactory factory = eventListenerFactories.get(name);
         checkArgument(factory != null, "Event listener factory '%s' is not registered. Available factories: %s", name, eventListenerFactories.keySet());
 
-        EventListener eventListener;
-        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(factory.getClass().getClassLoader())) {
-            eventListener = factory.create(properties);
-        }
+        EventListener eventListener = withClassLoader(factory.getClass(),
+                () -> factory.create(properties));
```
[the remark this answers is not in the document]
Comment (sergey-melnychuk) [reply core/trino-main/src/main/java/io/trino/eventlistener/EventListenerManager.java]: done: bda553f7574a15a4a3eb40f5c5ee5ea4ae8f3a0f

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java (hunk 10 above)
Comment (sergey-melnychuk) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java]: done: 248144248448ea297f075c119f15e4e1a83eeced

File: core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (sergey-melnychuk) [reply core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java:363]: done: 248144248448ea297f075c119f15e4e1a83eeced

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 14]
```diff
@@ -13,9 +13,24 @@
  */
 package io.trino.spi.classloader;
 
+import java.io.Closeable;
+
 public class ThreadContextClassLoader
+        implements Closeable
 {
-    private ThreadContextClassLoader() {}
```
Comment (kokosing) [inline core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java]: Please see https://github.com/trinodb/trino/pull/8411/files, we don't want one PR to remove something and the restore it. These two operations should be squashed. Simply don't remove it in a first place.

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 15]
```diff
@@ -13,9 +13,24 @@
...
+    private final ClassLoader originalThreadContextClassLoader;
+
+    public ThreadContextClassLoader(ClassLoader newThreadContextClassLoader)
```
Comment (kokosing) [inline core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java]: Please mark it as `@Deprecated` and write what should be used instead.

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java (hunk 14 above)
Comment (kokosing) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java]: Please apply this comment to other changes like that.

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java (hunk 15 above)
Comment (wendigo) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java]: Or maybe we could hide constructor here?

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java (hunk 14 above)
Comment (sergey-melnychuk) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java]: done, squashed everything into a single commit

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java (hunk 15 above)
Comment (sergey-melnychuk) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:27]: added `@Deprecated` annotation

Comment (sergey-melnychuk) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:27]: @wendigo How to hide the constructor? Any change to access modifiers would be breaking change of the SPI.

Comment (wendigo) [reply core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:27]: @sergey-melnychuk you are right :)

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 16]
```diff
@@ -31,4 +35,46 @@ public void close()
...
+    {
+        withClassLoader(newThreadContextClassLoader, () -> {
+            runnable.run();
+            return null;
+        });
+    }
+
+    public static <T, E extends Exception> T withClassLoader(ClassLoader newThreadContextClassLoader, ThrowingSupplier<T, E> supplier)
+    {
+        final ClassLoader originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
```
Comment (electrum) [inline core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:59]: We don’t use final on local variables

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 17]
```diff
@@ -31,4 +35,46 @@ public void close()
...
+        Thread.currentThread().setContextClassLoader(newThreadContextClassLoader);
+        try {
+            return supplier.get();
+        }
+        catch (RuntimeException e) {
+            throw e;
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
```
Comment (electrum) [inline core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:68]: This is wrong and different behavior than try-with-resources. With the generic parameter for the exception type, you can rethrow the originally. We actually don’t need catch at all — try/finally is sufficient.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/procedure/UnregisterPartitionProcedure.java [hunk 18]
```diff
@@ -84,9 +84,7 @@ public Procedure get()
 
     public void unregisterPartition(ConnectorSession session, ConnectorAccessControl accessControl, String schemaName, String tableName, List<String> partitionColumn, List<String> partitionValues)
     {
-        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
-            doUnregisterPartition(session, accessControl, schemaName, tableName, partitionColumn, partitionValues);
-        }
+        withClassLoaderOf(getClass(), () -> doUnregisterPartition(session, accessControl, schemaName, tableName, partitionColumn, partitionValues));
```
Comment (electrum) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/procedure/UnregisterPartitionProcedure.java:87]: Wrap

File: core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java [hunk 19]
```diff
@@ -337,12 +337,14 @@ private synchronized void removeConnectorInternal(CatalogName catalogName)
         MaterializedConnector materializedConnector = connectors.remove(catalogName);
         if (materializedConnector != null) {
             Connector connector = materializedConnector.getConnector();
-            try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(connector.getClass().getClassLoader())) {
-                connector.shutdown();
-            }
-            catch (Throwable t) {
-                log.error(t, "Error shutting down connector: %s", catalogName);
-            }
+            withClassLoaderOf(connector.getClass(), () -> {
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java:340]: This is worse than the original. Let’s revert this change and don’t deprecate the constructor version.

File: core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java [hunk 20]
```diff
@@ -165,8 +165,8 @@ public synchronized void stop()
 
         for (Map.Entry<CatalogName, MaterializedConnector> entry : connectors.entrySet()) {
             Connector connector = entry.getValue().getConnector();
-            try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(connector.getClass().getClassLoader())) {
-                connector.shutdown();
+            try {
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/connector/ConnectorManager.java:168]: Revert this one

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 21]
```diff
@@ -31,4 +35,46 @@ public void close()
     {
         Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
     }
+
+    public static void withClassLoaderOf(Class<?> clazz, Runnable runnable)
```
Comment (electrum) [inline core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:39]: I don’t think these “of” methods are used frequently enough to justify the API weight. Doing it in the caller seems more clear and isn’t much longer.

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 22]
```diff
@@ -31,4 +35,46 @@ public void close()
...
+        }
+        finally {
+            Thread.currentThread().setContextClassLoader(originalThreadContextClassLoader);
+        }
+    }
+
+    @FunctionalInterface
+    public interface ThrowingSupplier<T, E extends Exception>
+    {
+        T get() throws E;
```
Comment (electrum) [inline core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:78]: Wrap

File: core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java [hunk 23]
```diff
@@ -31,4 +35,46 @@ public void close()
...
+        withClassLoader(newThreadContextClassLoader, () -> {
+            runnable.run();
+            return null;
+        });
+    }
+
+    public static <T, E extends Exception> T withClassLoader(ClassLoader newThreadContextClassLoader, ThrowingSupplier<T, E> supplier)
+    {
+        final ClassLoader originalThreadContextClassLoader = Thread.currentThread().getContextClassLoader();
+        Thread.currentThread().setContextClassLoader(newThreadContextClassLoader);
```
Comment (electrum) [inline core/trino-spi/src/main/java/io/trino/spi/classloader/ThreadContextClassLoader.java:60]: Use try-with-resources

File: plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java [hunk 24]
```diff
@@ -98,13 +98,15 @@ void invalidateCache()
     @Override
     public Principal createAuthenticatedPrincipal(String user, String password)
     {
-        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
```
Comment (electrum) [inline plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapAuthenticator.java:101]: Revert

File: plugin/trino-raptor-legacy/src/test/java/io/trino/plugin/raptor/legacy/storage/TestShardWriter.java [hunk 25]
```diff
@@ -98,10 +98,11 @@ public void testWriter()
                 .row(null, "world", null, Double.POSITIVE_INFINITY, null, arrayBlockOf(BIGINT, 3, null), mapBlockOf(createVarcharType(5), BOOLEAN, "k2", null), arrayBlockOf(arrayType, null, arrayBlockOf(BIGINT, 6, 7)))
                 .row(456L, "bye \u2603", wrappedBuffer(bytes3), Double.NaN, false, arrayBlockOf(BIGINT), mapBlockOf(createVarcharType(5), BOOLEAN, "k3", false), arrayBlockOf(arrayType, arrayBlockOf(BIGINT)));
 
-        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(new EmptyClassLoader());
-                OrcFileWriter writer = new OrcFileWriter(columnIds, columnTypes, file)) {
-            writer.appendPages(rowPagesBuilder.build());
-        }
+        withClassLoader(new EmptyClassLoader(), () -> {
```
Comment (electrum) [inline plugin/trino-raptor-legacy/src/test/java/io/trino/plugin/raptor/legacy/storage/TestShardWriter.java:101]: Revert

**Review by electrum — CHANGES_REQUESTED** [review]
Generally looks good, and is a nice reduction of boilerplate, especially for the parts inside a method that need to assign a variable.

**Comment by colebow** [comment]
:wave: @sergey-melnychuk - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.
